### ISSUE TITLE
Copy and apply a gate

### ISSUE THREAD

**Issue by Mieszkol** [DillonHammill/CytoExploreR#74]
Hi Dillon, 
Would you consider adding a "copy&apply a gate" function? It'd be really useful for me - when comparing many populations by frequency of some antigen expression it'd be great if I could set the gate just once and then copy it onto another parent gates. Thank you!

Best
Mieszko

**Comment by DillonHammill** [comment]
@Mieszkol, looks like I added a function for this a while back but it is not currently exported for use. I will test it out today and if it works as expected I will export it for you.
https://github.com/DillonHammill/CytoExploreR/blob/720245f70b37750f1e420a6ec135edaa0e3d4cee/R/cyto_gate-helpers.R#L253-L355

**Comment by Mieszkol** [comment]
@DillonHammill That's great, thank you!

**Comment by Mieszkol** [comment]
@DillonHammill Any update on the function? Is it working or does it need rewriting?

**Comment by DillonHammill** [comment]
Sorry @Mieszkol, I thought I had commented on this again. Anyway, the function works, but I haven't pushed it yet as it doesn't have an editing mechanism at the moment. These type of (reference) gates operate differently to traditional gates and so their entries have to be updated directly in the gatingTemplate. The problem is that `cyto_gatingTemplate_edit()` is not currently working but will be working next week when I make the switch over to using `DataEditR`. So I am a bit hesitant in exporting this until this changeover has been made. Once everything is ready I will add a new issue documenting its use, as there are some things to keep in mind when using this feature.

**Comment by Mieszkol** [comment]
@DillonHammill Hi again Dillon, how did the swithover work out? When do you anticipate the copy gate function will be ready? I'm sorry for bothering you but I'm in great need of that feature

**Comment by Mieszkol** [comment]
@DillonHammill Hello again, is there any chance to see this function added rather soon? I'm in the final process of data preparation for the paper and unfortunately, without copy&apply I'll have to use other software for FCS analysis

**Comment by DillonHammill** [comment]
@Mieszkol I have exported the function for you. See below:
```{r}
# Copy CD69+ gate from CD4 T Cells to CD8 T Cells
cyto_gate_copy(gs, 
               parent = "CD8 T Cells", # parent of new population
               alias = "CD69+ CD8 T Cells", # new population name
               copy = "CD69+", # population to copy
               gatingTemplate = "gatingTemplate.csv")
```

**Comment by Mieszkol** [comment]
Thank you Dillon! Unfortunately, I can't make it work. I've tried to copy the gate POI+ to the CD56bright population using the the following command:

`cyto_gate_copy(gs, parent = "CD56bright", alias = "CD56bright POI+", copy = "POI+", gatingTemplate = gT)
`
and got this:
`Error in cyto_gate_copy(gs, parent = "CD56bright", alias = "CD56brightPOI+",  : 
  Supplied alias does not exist in the GatingSet.`

My cyto_nodes() output:
`> cyto_nodes(gs)
 [1] "root"                                         "/Cells"                                      
 [3] "/Cells/Single Cells"                          "/Cells/Single Cells/NK"                      
 [5] "/Cells/Single Cells/NK/CD56bright"            "/Cells/Single Cells/NK/CD56dim"              
 [7] "/Cells/Single Cells/NK/CD56dim/KIR2DL1+"      "/Cells/Single Cells/NK/CD56dim/KIR2DL1-"     
 [9] "/Cells/Single Cells/NK/CD56dim/KIR2DL2_3+"    "/Cells/Single Cells/NK/CD56dim/KIR2DL2_3-"   
[11] "/Cells/Single Cells/NK/CD56dim/KIR2DL1_S1+"   "/Cells/Single Cells/NK/CD56dim/KIR2DL1_S1-"  
[13] "/Cells/Single Cells/NK/CD56dim/KIR3DL1+"      "/Cells/Single Cells/NK/CD56dim/KIR3DL1-"     
[15] "/Cells/Single Cells/NK/CD56dim/NKG2A+ CD57-"  "/Cells/Single Cells/NK/CD56dim/NKG2A+ CD57+" 
[17] "/Cells/Single Cells/NK/CD56dim/NKG2A- CD57-"  "/Cells/Single Cells/NK/CD56dim/NKG2A-, CD57+"
[19] "/Cells/Single Cells/NK/CD56dim/CD57+, NKG2C+" "/Cells/Single Cells/NK/CD56dim/NKG2C-"       
[21] "/Cells/Single Cells/NK/CD56dim/NKG2A--"       "/Cells/Single Cells/NK/CD56dim/Single-KIR+"  
[23] "/Cells/Single Cells/NK/CD56dim/KIR-"          "/Cells/Single Cells/NK/CD56dim/2DL1+"        
[25] "/Cells/Single Cells/NK/CD56dim/2DL3+"         "/Cells/Single Cells/NK/CD56dim/Double KIR+"  
[27] "/Bool Ed KIR-"                                "/Bool Ed 2DL1+"                              
[29] "/Bool Ed 2DL3+"                               "/Bool Ed Double KIR+"                        
[31] "/Bool KIR-"                                   "/Bool KIR+"                                  
[33] "/CD56dim+ NKG2A+ CD57- NKG2C- KIR-"           "/CD56dim+ NKG2A+ CD57- NKG2C- KIR+"          
[35] "/CD56dim+ NKG2A- CD57- NKG2C- KIR+"           "/CD56dim+ NKG2A- CD57+ NKG2C- KIR2DL1+"      
[37] "/CD56dim+ NKG2A- CD57+ NKG2C- KIR2DL3+"       "/CD56dim+ NKG2A- CD57+ NKG2C+ KIR2DL1+"      
[39] "/CD56dim+ NKG2A- CD57+ NKG2C+ KIR2DL3+"       "/CD56dim+ NKG2A- CD57- NKG2C- KIR-"          
[41] "/CD56dim+ NKG2A- CD57- NKG2C- KIR2DL1+"       "/CD56dim+ NKG2A- CD57- NKG2C- KIR2DL3+"      
[43] "/CD56dim+ NKG2A- CD57- NKG2C- Double KIR+"    "/Cells/Single Cells/NK/POI+"  `

**Comment by Mieszkol** [comment]
Different error this time (nodes and eveything else is the same)

```
> cyto_gate_copy(gs, parent = "CD56bright", alias = "CD56bright POI+", copy = "POI+", gatingTemplate = gT)
Error in .gatingTemplate(dt, name = name) : 
  No dimensions defined for refGate!
```

**Comment by Mieszkol** [comment]
Hi @DillonHammill, have you had some more time to look at this issue?

**Comment by DillonHammill** [comment]
Sorry @Mieszkol, I didn't see your previous comment. I thought that the gate dimensions would be handled internally by `gs_add_gating_method` for reference gates, but I guess this is not the case. I have pushed a fix to include the gate dimensions in the gatingTemplate.
```{r}
devtools::install_github("DillonHammill/CytoExploreR")
```
I am currently away on vacation and have limited resources to test and debug  this, if things still don't work I will have to sort this out when  I return to work in January.

**Comment by Mieszkol** [comment]
Hi @DillonHammill, I hope you had a relaxing holiday break. 

I tried copy&apply after your fix but another error popped up:

`> cyto_gate_copy(gs, parent = "CD56bright", alias = "CD56bright POI+", copy = "POI+", gatingTemplate = gT)

Error in (function (classes, fdef, mtable)  : 
  unable to find an inherited method for function ‘parameters’ for signature ‘"list"’`

**Comment by DillonHammill** [comment]
That's strange, for some reason the previous commit does not match my local commit. Anyways, I just pushed a fix to the `master` branch:
```{r}
devtools::install_github("DillonHammill/CytoExploreR")
```

**Comment by Mieszkol** [comment]
It works, thank you so much! I'll test the function a bit more tomorrow and let you know if there are some issues.

**Comment by Mieszkol** [comment]
I've incorporated the function into my workflow and it seems to work without any issues. The only thing is this warning:

```
> cyto_gate_copy(gs, parent = "CD56bright", alias = "CD56bright POI+", copy = "POI+", gatingTemplate = gT)
A GatingSet with 9 samples
Warning messages:
1: In grepl(pname, pd$name, ignore.case = T) :
  argument 'pattern' has length > 1 and only the first element will be used
2: In grepl(pname, pd$name, ignore.case = T) :
  argument 'pattern' has length > 1 and only the first element will be used
3: In grepl(pname, pd$name, ignore.case = T) :
  argument 'pattern' has length > 1 and only the first element will be used
4: In grepl(pname, pd$name, ignore.case = T) :
  argument 'pattern' has length > 1 and only the first element will be used
5: In grepl(pname, pd$name, ignore.case = T) :
  argument 'pattern' has length > 1 and only the first element will be used
6: In grepl(pname, pd$name, ignore.case = T) :
  argument 'pattern' has length > 1 and only the first element will be used
7: In grepl(pname, pd$name, ignore.case = T) :
  argument 'pattern' has length > 1 and only the first element will be used
8: In grepl(pname, pd$name, ignore.case = T) :
  argument 'pattern' has length > 1 and only the first element will be used
9: In grepl(pname, pd$name, ignore.case = T) :
  argument 'pattern' has length > 1 and only the first element will be used
```

Anyway, thank you so much for adding this feature!
